allow querying connection metadata and then subsequent data in relayStylePagination

anark opened the pull request:
Currently if you try to query relay connection metadata with no arguments

```
query todos{
  totalCount
}
```

And then query for the edges/data
```
query todos{
  totalCount
  edges {
    node {
      title
    }
  }
}
```

The cache will never have any todo items and both ```data``` and ```error``` returned by the cache will always be undefined.

This attempts to fix this.
<!--
  Thanks for filing a pull request on Apollo Client!

  A few automated bots may chime in on your PR. They are here to help
  with reviewing and ensuring Apollo Client is production ready after each
  pull request merge.

    - apollo-cla will respond asking you to sign the CLA if this is your first PR.
      It may also respond with warnings, messages, or fail the build if something is off.
      Don't worry, it'll help you to fix what is broken!

    - bundlesize is a status check to keep the footprint of Apollo Client as small as possible.

    - circleci will run tests, checking style of code, and generally make
      sure everything is working as expected

  Please look at the following checklist to ensure that your PR
  can be accepted quickly:
-->

### Checklist:

- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests

apollo-cla commented:
@anark: Thank you for submitting a pull request! Before we can merge it, you'll need to sign the Apollo Contributor License Agreement here: https://contribute.apollographql.com/

benjamn commented on src/utilities/policies/pagination.ts:
```diff
@@ -121,7 +121,7 @@ export function relayStylePagination<TNode = Reference>(
 
       const updatePageInfo = (name: keyof TInternalRelay<TNode>["pageInfo"]) => {
-        const value = incoming.pageInfo[name];
+        const value = incoming.pageInfo && incoming.pageInfo[name];
```
This is a good idea. 👍

benjamn commented on src/utilities/policies/pagination.ts:
```diff
@@ -79,9 +79,9 @@ export function relayStylePagination<TNode = Reference>(
+      if (!args) return {...existing, ...incoming};
 
-      const incomingEdges = incoming.edges.slice(0);
+      const incomingEdges = (incoming.edges || []).slice(0);
```
This is a good idea. 👍

benjamn commented on src/utilities/policies/pagination.ts:
```diff
@@ -79,9 +79,9 @@ export function relayStylePagination<TNode = Reference>(
 
     merge(existing = makeEmptyData(), incoming, { args }) {
-      if (!args) return existing; // TODO Maybe throw?
+      if (!args) return {...existing, ...incoming};
```
Instead of combining `existing` with `incoming` like this (which could completely overwrite `existing.pageInfo`), do you think we could move the `args` check down to where `args` are used below, so that we skip over the `args.{after,before}` logic when `!args`, and just fall through to the normal logic for combining `existing` with `incoming`?

anark replied:
@benjamn Thanks for the feedback.

I've adjusted the implementation and extended the test coverage a bit.  Please let me know what else this needs.

anark commented:
@benjamn Thanks for the feedback.

I've adjusted the implementation, extended the test coverage to cover this, and updated from main. Please let me know what else this needs.

Ilaiwi commented:
I am dealing with an error that this PR could resolve here.
```
pagination.js:66 Uncaught TypeError: Cannot read property 'slice' of undefined
    at merge (pagination.js:66)
    at Policies.applyMerges (policies.js:294)
    at policies.js:316
```
I am getting this when I do a subscribe to more on a field with relay style pagination policy and attempting to update the cache where `existing` is passed as undefined to the `merge` function in the policy

benjamn reviewed (approved):
LGTM—thanks @anark!
